File: ABC/<200/180/C++/E.cpp
#include<iostream>
#include<vector>
#define rep(i,s,n) for(int i=s; i<(n); ++i)
#define INF 1e9
using namespace std;

int main(){
    int N; cin >> N;
    vector<vector<int>> XYZ(N, vector<int>(3, 0));
    rep(i,0,N)rep(j,0,3) cin >> XYZ[i][j];

    int maxS = (1<<N);
    vector<vector<int>> DP(N, vector<int>(maxS, INF));
    DP[0][0] = 0;
    rep(h,1,maxS)rep(cs,0,N)if(h&(1<<cs)){
        rep(ps,0,N){
            int d = abs(XYZ[cs][0]-XYZ[ps][0]) + abs(XYZ[cs][1]-XYZ[ps][1]) + max(0, XYZ[cs][2]-XYZ[ps][2]);
            DP[cs][h] = min(DP[cs][h], DP[ps][h-(1<<cs)]+d);
        }
    }
    cout << DP[0][maxS-1] << endl;
    return 0;
}

File: ABC/<200/180/Julia/E.jl
function localCost(p1::Array{Int}, p2::Array{Int})
    return abs(p2[1]-p1[1]) + abs(p2[2]-p1[2]) + max(0, p2[3]-p1[3])
end
function bitDP(M::Matrix{Int})
    R  = size(M)[2] # num of rows = num of site
    mS = 2^size(M)[2] # maximum number of the bit-expression of "sites that have been visited"
    DP = Inf.*ones(Float64, R, mS) # DP[current site, sites that have beem visited]
    DP[1,1] = 0
    for h ∈ 1:mS # h for histroy
        for cs ∈ 1:R # cr for current site
            if DP[cs,h] ≥ Inf
                continue
            end
            for ns ∈ 1:R # ns for next site
                if ns == cs || ((h-1)÷2^(ns-1))%2 == 1 # ns is already visited
                    continue
                end
                d = localCost(M[:, cs], M[:,ns])
                DP[ns, h+2^(ns-1)] = min(DP[ns, h+2^(ns-1)], DP[cs, h]+d)
            end
        end
    end
    return Int(floor(DP[1, mS]))
end

function main()
    N = parse(Int, readline())
    # Sites
    S = Array{Int}(undef, 3, N)
    for c ∈ 1:N
        S[:,c] = parse.(Int, readline()|>split)
    end
    # show
    bitDP(S) |> println
end
main()